src/java.base/share/classes/sun/security/ssl/X509KeyManagerImpl.java

```diff
@@ -77,15 +79,58 @@ final class X509KeyManagerImpl extends X509ExtendedKeyManager
     // cached entries
     private final Map<String,Reference<PrivateKeyEntry>> entryCacheMap;
 
+    private final boolean ksP12;
+    /*
+     * The credentials from the PKCS12KeyStore as
+     * Map: String(builderIndex.entryAlias) -> X509Credentials(credentials)
+     */
+    private final Map<String, X509Credentials> credentialsMap;
+
+    private static class X509Credentials {
+        final PrivateKey privateKey;
+        final X509Certificate[] certificates;
+
+        X509Credentials(PrivateKey privateKey, X509Certificate[] certificates) {
+            this.privateKey = privateKey;
+            this.certificates = certificates;
+        }
+    }
+
     X509KeyManagerImpl(Builder builder) {
         this(Collections.singletonList(builder));
     }
 
     X509KeyManagerImpl(List<Builder> builders) {
         this.builders = builders;
         uidCounter = new AtomicLong();
-        entryCacheMap = Collections.synchronizedMap
-                        (new SizedMap<>());
+        boolean foundPKCS12 = false;
+
+        for (Builder builder : builders) {
+            try {
+                KeyStore keyStore = builder.getKeyStore();
+                String ksType = keyStore.getType();
+                if (ksType.equalsIgnoreCase("pkcs12")) {
+                    foundPKCS12 = true;
```

Comment:
this will enable caching if _any_ of the keystores is PKCS12; we should only enable caching when _all_ keystores are PKCS12